### ISSUE TITLE
Bad memory leak on FreeBSD

### ISSUE THREAD

**Issue by varosi** [snoyberg/http-client#322]
This is a memory leak that is critical for us.
We catched it and all the profiling files are in the archive:
[git.zip](https://github.com/snoyberg/http-client/files/1686219/git.zip)

I hope this could help!
Application is running for long time and it was taking 32Gb RAM! We have constrained it to 1Gb heap to catch the problem. We catched it on the server. I though that memory leak was in our code, but it actually follow to network library as I see profiling info.

Could someone help with that problem? We could help with more debugging if needed. We have a way to reproduce it now, so if someone see it soon we could help better.

OS is "FreeBSD 11.1-STABLE". Stack LTS is: 10.3

**Comment by snoyberg** [comment]
I had some trouble generating the `ps` file with `hp2ps`:

```
hp2ps: processJira.log.hp, line 5110: integer must follow identifier
```

Deleting that line gave me a graph, but it was not what I expected:

<img width="862" alt="screen shot 2018-02-01 at 8 44 44 pm" src="https://user-images.githubusercontent.com/49415/35696677-c32173ee-0790-11e8-9151-b7749dc3ec82.png">

From what I can tell, it looks like the memory usage is coming from the `tls` library. Can you confirm that you're seeing something similar in your profiling info?

CC @vincenthz in case you want to take a look at this.

**Comment by ocheron** [comment]
A large file left in the system certificate directory could explain this (i.e. in /usr/local/share/certs).

**Comment by vincenthz** [comment]
It's very likely has to do with the way we load all certificates in memory at startup. it would be nice to have a more on-demand way, but that requires some significant changes to the way it works currently. I agree with @ocheron, anything funky in the system certificate could trigger this

**Comment by varosi** [comment]
@snoyberg I even cannot open .ps file from Ghost Viewer 5 on Windows.

**Comment by ocheron** [comment]
No I don't think 798K can explain the gigabytes, even with 50 managers being initialized in parallel.

Another quick question: in `src` you seem to have local copies of some packages like `cereal`. Is this compiled with optimizations?

**Comment by varosi** [comment]
This file contains mainly Mozilla certificates and it's very usual in size.
Options to build the executable are:

> -Wall
> -O2
> -threaded
> -prof
> -fprof-auto
> -rtsopts
> -fprof-cafs

Dependencies are:

> base >= 4.6 && < 5.0,
> aeson,
> bytestring,
> text,
> http-types,
> http-conduit,
> http-client,
> http-client-tls,
> process,
> process-extras,
> attoparsec

**Comment by varosi** [comment]
@ocheron, here is the normal execution log file:

[ProcessJIRA_logs_normal.zip](https://github.com/snoyberg/http-client/files/1698989/ProcessJIRA_logs_normal.zip)

**Comment by varosi** [comment]
Here I successfully get correct profiling files with another reproducing of this memory leak:
[secondLeak.zip](https://github.com/snoyberg/http-client/files/1699945/secondLeak.zip)

![image](https://user-images.githubusercontent.com/6051718/35872065-850d95ce-0b6e-11e8-81c3-e3c27dd44a6b.png)

**Comment by varosi** [comment]
Data.X509.CertificateStore.readCertificateStore take almost all the time for some reason.

**Comment by varosi** [comment]
@vincenthz is it possible to use pemParseBS instead of pemParseLBS, so it's strictly parsed/readed?

**Comment by ocheron** [comment]
If all profiling reports are on the same machine, then I don't see performance issue reading the certificate store when total execution is 0.37s.

Yes `decodeLenient` is kind of slow, like mentioned last year in vincenthz/hs-pem#5, but this does not explain why your program retains memory.

And if I trust what I see in the second report, the explanation may be simply like this: each of 50 calls to `getHTTP` calls `newManager`, and the result from `httpLbs` is probably not used strictly enough.
So a lot is retained, including 50 copies of the certificate store.

If correct here, priority should be to use only one manager for all queries. Then effect of laziness will be much less.

**Comment by varosi** [comment]
You're right that there is too much lazyness, but it goes long after the hs-certificate library as an effect. [readCertificates](https://github.com/vincenthz/hs-certificate/blob/207d9fef631234df4303c9acb1519a2e4241b26f/x509-store/Data/X509/CertificateStore.hs#L77) function is reading a file from the file system as IO operation for certificates. Isn't it good idea if that function is strictier as proposed in [this pull request](https://github.com/vincenthz/hs-certificate/pull/93)?

**Comment by ocheron** [comment]
And did you manage to fix the issue?

Regarding the PR I don't think it can improve anything:
- forcing a string ByteString like `pemContent` is a no-op
- other elements are not used and discarded, so there is no need to evaluate either

**Comment by varosi** [comment]
@ocheron, PR didn't stopped it from leaking. So you could ignore it. Currently we solved the problem with not initializing network manager more than once. But I'm curious why it is so slow? Usually between 300-500ms on a modern machine which is way too slow.

**Comment by ocheron** [comment]
With pem >= 0.2.3 it should be faster (reason already explained in what I referenced above).
